### PR TITLE
Logic changes so that all items can, in principle, be placed in their vanilla locations

### PR DESCRIPTION
**Pull request by MrCheeze**
One possible sanity check of the randomization logic if making sure that "absolutely everything vanilla' is a possible result. Currently there are a few reasons why such a result is impossible, this pull request changes that.

Changes that apply always (keysanity ON or off):

* Change Cake from replenishable to non-replenishable, and change Gourmet Guy to expect you to cook a Cake Mix instead of giving him the non-replenishable cake. Reasoning: this is the "logic" that the original game uses.

Changes that only apply when key items are restricted to their original dungeon (keysanity OFF):

* Assume the player has Super Boots when filling items for Tubba Blubba's Castle, Flower Fields, and Crystal Palace. (Currently keys are placed only in places the super boots don't block off.)

* In Toy Box, assume that if the player has the Toy Train, they are able to throw it inside, that if they have the Cake Mix they can cook it, and that if they have the Dictionary+Mystery Note they can translate it. (Currently keys are all frontloaded at blue station.)

* Treat the 4 items in the Toad Town Storeroom as if they were inside Toy Box, meaning that keys such as the Toy Train can be randomized there. (I didn't give any other Toad Town checks are given this treatment though, not even the rewards for returning Dictionary/Calculator/Mailbag/FryingPan, but that's something to consider too.)

Everything here should be taken more as me demonstrating suggestions than anything, and is worth discussing, especially the storeroom change.

### CODE REVIEW

File: maps/graph_edges/edges_omo.py [hunk 1]
```diff
@@ -110,7 +110,7 @@
     {"from": {"map": "OMO_05", "id": 1},               "to": {"map": "OMO_05", "id": "HiddenYBlockB"}, "reqs": [require(partner="Watt",flag="RF_HiddenBlocksVisible")]}, #* PNK Gourmet Guy Crossing Exit East (North) -> HiddenYBlockB (SuperSoda)
     {"from": {"map": "OMO_05", "id": "HiddenYBlockB"}, "to": {"map": "OMO_05", "id": 1},               "reqs": []}, #* HiddenYBlockB (SuperSoda) -> PNK Gourmet Guy Crossing Exit East (North)
 
-    {"from": {"map": "OMO_05", "id": 0}, "to": {"map": "OMO_05", "id": 0}, "reqs": [require(item="Cake")], "pseudoitems": ["MF_Ch4_GaveCakeToGourmetGuy"]}, #+ Give Cake to Gourmet Guy
+    {"from": {"map": "OMO_05", "id": 0}, "to": {"map": "OMO_05", "id": 0}, "reqs": [require(flag="RF_CanVisitTayceT"),require(item="CakeMix")], "pseudoitems": ["MF_Ch4_GaveCakeToGourmetGuy"]}, #+ Give Cake to Gourmet Guy
```
Comment (Pronyo-Chan) [inline maps/graph_edges/edges_omo.py:113]: Maybe the requirement should be CakeMix OR Cake? Could even include the other cakes although that would be pretty irelevant considering the low chance of them appearing in most seeds

Comment (MrCheeze) [reply maps/graph_edges/edges_omo.py:113]: Possibly. Generation wouldn't actually be affected, though, unless Cake was added back into `progression_miscitems` too. (And adding it to progression_miscitems would make it impossible for the Cake to appear in a non-replenishable location such as the vanilla location, I believe.)

**Comment by MrCheeze** [comment]
Another possibility for the keysanity-off Toy Train would to simply allow it to be placed anywhere in the world (and assume it's obtained when placing items in Toy Box).

File: maps/graph_edges/edges_omo.py (hunk 1 above)
Comment (icebound777) [reply maps/graph_edges/edges_omo.py:113]: We have to have one Cake in the pool anyway, as giving TayceT. her FryingPan immediately gives the Cake with no need for ever getting a CakeMix.
Which is also why I'm a bit meh on the logic change for GG, as I'm pretty sure I've never used a CakeMix in casual play for making a Cake, instead I always handed in the one I got in exchange for the FryingPan.
Though I can see changing logic to (Cake OR (CakeMix AND TayceTReachable)).

**Comment by MrCheeze** [comment]
Updated after discussion to revert the changes to Cake (so it still gets placed in a renewable location, and can't be placed in a one-time check like it is in the vanilla game) and to the toy train (so that it still appears in toy box and can't appear in the storeroom with keysanity off).

File: rando_modules/logic.py [hunk 2]
```diff
@@ -540,10 +547,9 @@ def _get_limit_items_to_dungeons(
                 False,
                 False
             )
-        add_to_inventory([
-            "CrystalBerry",
-            "WaterStone"
-        ])
+        if area_name in additional_starting_items:
+            for item in additional_starting_items[area_name]:
```
Comment (icebound777) [inline rando_modules/logic.py:551]:
```suggestion
            add_to_inventory(additional_starting_items[area_name])
```  
and remove line below.
Reasoning: add_to_inventory has handling for an argument of type list

**Review by icebound777 — APPROVED** [review]
Looks fine to me.  
I'll adjust the CakeMix / Cake logic later, as mentioned on discord.
Might want to add handling for a failing assert during the check for chapter beatability some time, but not strictly necessary

**Pull request merged**